php/libraries/NDB_Page.class.inc

```diff
@@ -590,6 +595,83 @@ class NDB_Page
         return $this->form->createElement('password', $field, $label, $attr);
     }
 
+    /**
+     * An NDB_Page can act as both a middleware, or a handler, depending on the
+     * context. When treated as a middleware, the page is responsible for setting
+     * up any middleware required to access the page. By default, this will
+     * implement the page decoration middlewares which is responsible for adding
+     * the LORIS menu's/headers/footers/etc.
+     * This can be overridden by modules who don't want this behaviour (such
+     * as API endpoints.)
+     *
+     * @param ServerRequestInterface  $request The PSR7 request being processed.
+     * @param RequestHandlerInterface $handler The handler to handle the request
+     *                                         after processing the middleware.
+     *
+     * @return ResponseInterface the PSR15 response that was generated by the
+     * middleware
+     */
+    public function process(
+        ServerRequestInterface $request,
+        RequestHandlerInterface $handler
+    ) : ResponseInterface {
+        $baseURL = $request->getAttribute("baseurl");
+        $config  = NDB_Config::singleton();
+
+        $user = $request->getAttribute("user");
+
+        // Add the page class to the request to that middleware that needs it
+        // has access to it. (This currently required by the page decoration
+        // middleware, because of the control panels.)
+        $request = $request->withAttribute("pageclass", $this);
+
+        // FIXME: For historical compatibility with NDB_Menu_Filter.
+        // Format should be deprecated and getting the page in different formats
+        // should be a different endpoint.
+        $get = $request->getQueryParams();
+        if (isset($get['format'])) {
+            $this->format = $get['format'];
+            // Do not add the decoration middleware for json format requests.
```

Comment:
[comment] => Also, there is content type negotiation. Pages that support/expect different format should handle it by checking the headers.

Reply:
I agree, that's what the FIXME comment above is getting at, but this is currently required for backwards compatibility with existing LORIS modules, otherwise every NDB_Menu_Filter page breaks.